### ISSUE TITLE
Call PostgreSQL function from Lizmap.

### ISSUE THREAD

**Issue by groschialeux** [3liz/lizmap-web-client#1444]
I am using Lizmap to create a front end for pgRouting.  The user will use Lizmap interface to add points to a table and then click a button to call a PostgreSQL function that calculates a route for those points and save it as a multilinestring in a table that is a layer in my project.

I figured out how to add the button and now I am looking for the best way of calling the function.

```
lizMap.events.on({
    uicreated: function(e) {
      lizMap.addDock(
            'routing',
            'Itinéraire',
            'minidock',
            '<input type="button" value="Calculer"> onclick="myRouting()"',
            'icon-road'
        );
        //-------------------------------------------
        $i = 0;
        $(function () {
          $('#mapmenu li.routing a').click(function () {
            if ($i === 0) {
              $i = 1;
            } else {
              $i = 0;
            }
            });
          });
    }
});
```

At first I was thinking of creating the myRouting() function in javascript with all the code needed to connect to the database, and call the function but I am wondering if it is the best way.  The function does not require any parameter and does not return anything meaningful.  I simply need my button to execute SQL command "select webmap.calculroute()".

Is there a function or something in Lizmap that could serve as a wrapper to pass a SQL command to PostgreSQL ?

**Comment by nboisteault** [comment]
I think you can use a trigger which refresh the multilinestring table when a point is added in first table. Then refresh tables with WMS/WFS requests.

**Comment by groschialeux** [comment]
I thought about a trigger but it is not the best solution.  The users will create multiple points before he or she is ready to calculate a route.

The calcroute() function takes about 20 seconds to run, there is no reason for calling it before the user is finished setting up all the waypoints.

My question really is on the "best practices" side of things.  I am really just asking if there is something ready in Lizmap to interact with postgresql.   If there is nothing then I will simply write my code in php and execute it when the user clicks the button.

**Comment by nboisteault** [comment]
WFS interacts with SQL.

**Comment by mdouchin** [comment]
We just added a new feature in lizmap web client called actions which allows to add action buttons in the popup 
https://github.com/3liz/lizmap-documentation/blob/master/source/publish/advanced_lizmap_config.rst#add-action-buttons-in-the-popup

There is no configuration interface yet in the plugin.
We also plan to make this feature more generic and powerfull in the future.

This feature will be available in lizmap web client 3.4

**Comment by groschialeux** [comment]
That's going to be a nice feature.   In the meantime I decided to use a trigger with a little trick.   I have a field where the user enters the waypoint step number.  That tells the order of the points to the routing logic.

I created my trigger so it calculates the route only where there is one waypoint with step number 999999 so the user can enable/disable route recalculation by simply editing the step number of his last waypoint.
